cpp/src/arrow/dataset/filter.h

```diff
@@ -377,6 +377,8 @@ class ARROW_DS_EXPORT CastExpression final
 
   const compute::CastOptions& options() const { return options_; }
 
+  std::shared_ptr<DataType> to_type() const;
```

Comment:
```suggestion
  /// Return the target type of this CastTo expression, or nullptr if this is a CastLike expression
  const std::shared_ptr<DataType>& to_type() const {
    if (arrow::util::holds_alternative<std::shared_ptr<DataType>>(to_)) {
      return arrow::util::get<std::shared_ptr<DataType>>(to_);
    }
    static std::shared_ptr<DataType> null;
    return null;
  }

  /// Return the target expression of this CastLike expression, or nullptr if this is a CastTo expression
  const std::shared_ptr<Expression>& like_expr() const {
    if (arrow::util::holds_alternative<std::shared_ptr<Expression>>(to_)) {
      return arrow::util::get<std::shared_ptr<Expression>>(to_);
    }
    static std::shared_ptr<Expression> null;
    return null;
  }
```

Reply:
Thanks! I'm not sure why is variant used for the cast target, but I guess it will be implemented differently in the next iteration of the expressions.